lib/kafka/producer.rb

```diff
@@ -218,6 +224,17 @@ def produce(value, key: nil, topic:, partition: nil, partition_key: nil)
         partition_key: partition_key,
       )
 
+      if buffer_size >= @max_buffer_size
+        raise BufferOverflow, "Max buffer size (#{@max_buffer_size} messages) exceeded"
+      end
+
+      if buffer_bytesize + message.bytesize >= @max_buffer_bytesize
```

Comment:
For the count limit, we know whether we'll exceed the limit if we're currently _at it_. For the byte size limit we need to see if adding the message would put us _over the limit_.